### DISCUSSION TITLE
JSI::Schema::ReferenceError - cannot find schema

### DISCUSSION THREAD

**Discussion by makemp** [notEthan/jsi#288] [Q&A] [answered]
I have a problem with `ref` marker in yaml.

I have code:
`JSI.new_schema(YAML.load_file(file_path))`
Which raises an error 

```
   JSI::Schema::ReferenceError:
       cannot find schema by ref: ./definitions.yml#/definitions/draft
       from: #{<JSI (JSI::JSONSchemaOrgDraft07) Schema>
         "$ref" => "./definitions.yml#/definitions/draft"
       }
```
Because it loads a file which looks like that:

```
$schema: 'http://json-schema.org/draft-07/schema'
description: 'desc goes here'
required: ['person']
properties:
  person:
    $ref: "./definitions.yml#/definitions/draft"
```
And `definitions.yml` file is in the same dir as 'file_path' file
and looks like that:

```
$schema: 'http://json-schema.org/draft-07/schema'
definitions:
  draft:
    type: object
    properties: ...
```

I do not know what I am doing wrong but It would be great to share definitions outside the file.
Any ideas?
Thaank you

**Comment by notEthan** [comment, marked as answer]
At the moment, JSI has somewhat limited capabilities when it comes to loading schemas, so it can be a little bit cumbersome on the application. I have plans for how to improve this, but implementing that is, as yet, behind other priorities.

JSI doesn't do any loading of schemas' content itself, from the filesystem or network - for every schema document, the application has to load its data, parse JSON or YAML, and pass that to `new_schema`.

From there, any relative URI in a `$ref` must be able to resolve to an absolute URI. (A mechanism to autoload some range of URIs from a filesystem or network, and instantiate those as schemas with appropriate absolute base URIs, is planned.)

So, a few things to address to get this going:

- The referenced schema document needs an absolute URI. You can declare this with `$id` (preferable), or you can use the `uri` param to `JSI.new_schema`, e.g. `JSI.new_schema(YAML.load("/path/definitions.yml"), uri: "file:/path/definitions.yml")`. Note that JSI doesn't use the protocol, the URI may be `file` or `http` or any valid URI.
- To resolve a `$ref`, its target needs to be registered. Instantiating with `new_schema` does register the schema, or the registry can autoload a URI lazily when referenced, something like:

  ```ruby
  JSI.schema_registry.autoload_uri("file:/path/definitions.yml") do
    JSI.new_schema(YAML.load("/path/definitions.yml"))
    # again, add a param like `uri: "file:/path/definitions.yml"` to `new_schema` if it has no `$id`
  end
  ```

  Those semantics / flexibility of autoloading is one of the major things I want to improve a lot.

- Since `$ref: "./definitions.yml#/..."` is relative, resolving that needs a base URI. The referencing schema (which I'll call `person`) should also have an id like `file:/path/person.yml`, something adjacent to the `definitions` schema's URI, specified the same way with `$id` schema keyword or `uri` param of `new_schema`.

In summary, the flow to create instances of the `person` schema generally looks like:
- For each URI that a `$ref` will point to, register that URI by either instantiating the schema containing it (with `new_schema`) or setting an `autoload_uri` to instantiate it
- Initialize the `person` schema with `new_schema`, with its own absolute URI
- Initialize instances of the `person` schema with `new_jsi` - references should resolve successfully.

**Reply by makemp** [reply]
Thank you. That helped a lot.
while `$ref` I am using `$id` and it works for me.
